src/snovault/elasticsearch/create_mapping.py

```diff
@@ -133,10 +132,12 @@ def schema_mapping(name, schema):
             'type': field_type
         }
 
-        # these fields are unintentially partially matching some small search
-        # keywords because fields are analyzed by nGram analyzer
-        if name in NON_SUBSTRING_FIELDS:
-            sub_mapping['include_in_all'] = False
+        if name not in NON_SUBSTRING_FIELDS:
+            if depth == 1 or (depth == 2 and parent == 'array'):
```

Comment:
I think this is the logic I'm most uncomfortable with, though the results certainly look closer than before. What's the implication of not tracking depth, as it didn't seem like it was tracked before? Is it possible to always copy to _all `if name not in NON_SUBSTRING_FIELDS`?

Reply:
Older demos showed that setting `"copy_to": "_all"` always here is equivalent to boosting every single field in the nested objects (minus the `NON_SUBSTRING_FIELDS`). Depth tracking stops `"copy_to": "_all"` after level 1. Before, if file or experiment award description has "brain" and "blood" in the text, the files or experiments show up under either "blood" or "brain" search term. Non-specific experiments that showed up in the results in earlier demos were due to "copy_to" being applied to the fields of deeply nested objects. Older `include_in_all` was applied by default and depth tracking is applied under the hypothesis that, this default application of the setting was weighed more towards top fields rather than nested object fields. Result is that, now all top fields of the object has `copy_to_all`, and nested fields need to be boosted like so `"file.award.description": 1.0`. We could have achieved the same matching results by boosting all of the individual fields of a type also without depth tracking, which is still an option but may take more effort to match results.

Reply:
It seems like maybe defining in boost would make sense given we already have top-level fields there, e.g.:
 ```
"boost_values": {
        "accession": 20.0,
        "@type": 1.0,
        "alternate_accessions": 1.0,
        "assay_term_name": 20.0,
        "assay_term_id": 1.0,
        "assay_title": 20.0,
        "assay_slims": 5.0,
        "dbxrefs": 1.0,
        "aliases": 1.0,
        "biosample_ontology.term_id": 1.0,
        "biosample_ontology.term_name": 10.0,
```

> this default application of the setting was weighed more towards top fields rather than nested object fields.

I don't think this explains the case of award.description with blood or brain. Even if the weight was low on these they would still show up in results, though at the bottom. However we're not seeing them at all on current prod.

My understanding is that if `_all` is enabled on old ES then `include_in_all` is True by default. However you can set it to `include_in_all: False` for a certain field and it will apply to all subfields unless otherwise specified. I think looking at new and old mappings informative here. For file:

<img width="524" alt="Screen Shot 2020-10-27 at 11 26 38 AM" src="https://user-images.githubusercontent.com/16053922/97346430-a92de000-1848-11eb-8f9a-8e6e505e5837.png">

<img width="535" alt="Screen Shot 2020-10-27 at 11 27 55 AM" src="https://user-images.githubusercontent.com/16053922/97346443-ad59fd80-1848-11eb-8688-7be157887f47.png">

I'm a little concerned since the default has changed from don't include unless specifically set to always include if in first level (probably explains why we are seeing more hits on new demo?). The boost offers more specific control, while always including top-level fields doesn't.

In any case might be close enough for our current purposes, and could be tweaked later. But I think worth looking more into.

Reply:
You are right that boost offers more specific control. I find that it's not feasible to determine what to boost for the number of types we have multiplied by the number of top fields for each type, make demo for each combination, and compare results with the old ES. `ctcf` search bringing up `File` type results was due to unintentional boost on the `aliases` field. Determining which other objects and fields were important to match the existing result count in the old ES maybe challenging.

Reply:
I think it's likely the difference is just due to not copying the `links` and `unique_keys` stuff to `_all`, which is by default `include_in_all: True` on old ES:
<img width="372" alt="Screen Shot 2020-10-27 at 4 02 44 PM" src="https://user-images.githubusercontent.com/16053922/97371690-0f2d5e00-186f-11eb-9915-ed462d200b39.png">

Pretty sure it was the alias in unique keys that was matching `ctcf`, not the the alias in embedded properties. (Can confirm that alias in embedded properties is False for old mapping.)

It doesn't look like any of the embedded properties in old mapping are `include_in_all: True` unless they are defined in boost values for that schema. Could we go back to only copying embedded properties to _all if they are defined in boost, and just make the dynamic template links/unique keys have copy_to?

> ctcf search bringing up File type results was due to unintentional boost on the aliases field

Yeah, looks like only field ctcf is mentioned in file is alias and submitted_file name, but it's probably fine to have these intentionally searchable.

Reply:
Here is another demo with the changes: https://elasticsearch7-without-depth-tracking.demo.encodedcc.org/